components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt

```diff
@@ -140,6 +140,13 @@ open class GleanInternalAPI internal constructor () {
         // This must be set before anything that might trigger the sending of pings.
         initialized = true
 
+        // There are now dispatcher tasks that we need to perform *before* any tasks
+        // that may have been queued prior to initialization.  Therefore, we turn off
+        // task queueing now, so we can set those early tasks up before actually
+        // flushing the queue below.
+        @Suppress("EXPERIMENTAL_API_USAGE")
+        Dispatchers.API.setTaskQueueing(false)
```

Comment:
I think that this may be dangerous as I think it opens up a case where we can get the tasks executed out of order, especially if there is enough time between this and flushing of the queued tasks.  

Since this is the _only_ place that `onReadyToSendPings` is called, maybe it would be better to change the implementation of that function to not use `Dispatchers.API.launch`. Instead create a unique coroutine scope for the purpose of `onReadyToSendPings` or perhaps use `Dispachers.IO`.  The only other coroutines that I could find that get triggered as a result of `onReadyToSendPings`are the ones that serialize the pings to disk if `sendPingsByName` is called. Fortunately, these coroutines aren't using our `API` couroutine scope but `Dispatchers.IO`.  That means that neither they, nor a refactored `onReadyToSendPings` would be subject to the queueing.

Reply:
> I think it opens up a case where we can get the tasks executed out of order, especially if there is enough time between this and flushing of the queued tasks.

That's a good point.  If other tasks get called between this line and the flushing, they will get run *before* everything in the queue.

If `onReadyToSendPings` is called on its own coroutine scope, then there is no guarantee that it will run before any event recording tasks already in the queue.

`onReadyToSendPings` calls `sendPingsByName` which is `sendPings` under the hood which *does* use `Dispatchers.API` for ping collection.  It's only the serialization and writing to disk that happens on the IO thread.  This ping collection needs to happen before any pre-queued tasks, or we risk co-mingling events from two different boot timers in the same ping.

So let me go back to the drawing board.  Obviously flipping the global switch on whether things go to the "real" queue or the secondary task queue that we manage is the wrong approach.  I would just add an API to say "queue this with the real dispatcher now and never put in the task queue", but that doesn't handle the secondary things that that also queues (the `sendPing` calls).  <sigh>